int-set/src/bitpage.rs

```diff
@@ -0,0 +1,353 @@
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    // TODO(garretrieger): iterator that starts after some value (similar to next in hb).
+    // TODO(garretrieger): reverse iterator.
+
+    /// Iterator over the members of this page.
+    pub fn iter(&self) -> impl Iterator<Item = u32> + '_ {
+        self.storage
+            .iter()
+            .enumerate()
+            .filter(|(_, elem)| **elem != 0)
+            .flat_map(|(i, elem)| {
+                let base = i as u32 * ELEM_BITS;
+                iter_bit_indices(*elem).map(move |idx| base + idx)
+            })
+    }
+
+    /// Marks (val % page width) a member of this set.
+    pub fn insert(&mut self, val: u32) -> bool {
+        let ret = !self.contains(val);
+        *self.element_mut(val) |= elem_index_bit_mask(val);
+        self.mark_dirty();
+        ret
+    }
+
+    /// Marks all values [first, last] as members of this set.
+    pub fn insert_range(&mut self, first: u32, last: u32) {
+        let first = first & PAGE_MASK;
+        let last = last & PAGE_MASK;
+        let first_elem_idx = first / ELEM_BITS;
+        let last_elem_idx = last / ELEM_BITS;
+
+        for elem_idx in first_elem_idx..=last_elem_idx {
+            let elem_start = first.max(elem_idx * ELEM_BITS) & ELEM_MASK;
+            let elem_last = last.min(((elem_idx + 1) * ELEM_BITS) - 1) & ELEM_MASK;
+
+            let end_shift = ELEM_BITS - elem_last - 1;
+            let mask = u64::MAX << (elem_start + end_shift);
+            let mask = mask >> end_shift;
+
+            self.storage[elem_idx as usize] |= mask;
+        }
+
+        self.mark_dirty();
+    }
+
+    /// Removes (val % page width) from this set.
+    pub fn remove(&mut self, val: u32) -> bool {
+        let ret = self.contains(val);
+        *self.element_mut(val) &= !elem_index_bit_mask(val);
+        self.mark_dirty();
+        ret
+    }
+
+    /// Return true if (val % page width) is a member of this set.
+    pub fn contains(&self, val: u32) -> bool {
+        (*self.element(val) & elem_index_bit_mask(val)) != 0
+    }
+
+    fn mark_dirty(&mut self) {
+        self.len.set(u32::MAX);
+    }
+
+    fn is_dirty(&self) -> bool {
+        self.len.get() == u32::MAX
+    }
+
+    fn element(&self, value: u32) -> &Element {
+        let idx = (value & PAGE_MASK) / ELEM_BITS;
+        &self.storage[idx as usize]
+    }
+
+    fn element_mut(&mut self, value: u32) -> &mut Element {
+        let idx = (value & PAGE_MASK) / ELEM_BITS;
+        &mut self.storage[idx as usize]
+    }
+}
+
+/// returns the bit to set in an element for this value
+const fn elem_index_bit_mask(value: u32) -> Element {
+    1 << (value & ELEM_MASK)
+}
+
+fn iter_bit_indices(val: Element) -> impl Iterator<Item = u32> {
+    let mut idx = 0;
+
+    std::iter::from_fn(move || {
+        if idx >= ELEM_BITS {
+            return None;
+        }
+        let mask = 1u64.checked_shl(idx).unwrap_or(0) - 1;
```

Comment:
since we know that `idx < ELEM_BITS` we I don't think we need the `checked_shl` (or the `unwrap_or`)?